oauthproxy.go

```diff
@@ -622,12 +625,52 @@ func (p *OAuthProxy) IsWhitelistedPath(path string) bool {
 	return false
 }
 
-func getRemoteAddr(req *http.Request) (s string) {
-	s = req.RemoteAddr
-	if req.Header.Get("X-Real-IP") != "" {
-		s += fmt.Sprintf(" (%q)", req.Header.Get("X-Real-IP"))
+// GetRealClientIP obtains the IP of the end user client from headers or remote address.
+// Returns (nil, nil) if no client IP found matching requirements found (missing header).
+func (p *OAuthProxy) GetRealClientIP(req *http.Request) (net.IP, error) {
+	if p.realClientIPParser != nil {
+		return p.realClientIPParser.GetRealClientIP(req.Header)
 	}
-	return
+	return p.GetRemoteIP(req)
+}
+
+// GetRemoteIP obtains the IP of the low-level connected network host
+func (p *OAuthProxy) GetRemoteIP(req *http.Request) (net.IP, error) {
+	if ipStr, _, err := net.SplitHostPort(req.RemoteAddr); err != nil {
+		return nil, fmt.Errorf("Unable to address (%s) from http.RemoteAddr", req.RemoteAddr)
+	} else if ip := net.ParseIP(ipStr); ip != nil {
+		return ip, nil
+	} else {
+		return nil, fmt.Errorf("Unable to parse IP (%s)", ipStr)
+	}
+}
+
+// GetClientString obtains the human readable string of the remote IP and optionally the real client IP if available
+func (p *OAuthProxy) GetClientString(req *http.Request, full bool) (s string) {
+	var realClientIPStr string
+	if p.realClientIPParser != nil {
+		if realClientIP, err := p.realClientIPParser.GetRealClientIP(req.Header); err != nil {
+			logger.Printf("Unable to get real client IP: %v", err)
+		} else if realClientIP != nil {
+			realClientIPStr = realClientIP.String()
+		}
+	}
+
+	var remoteIPStr string
+	if remoteIP, err := p.GetRemoteIP(req); err == nil {
+		remoteIPStr = remoteIP.String()
+	} else {
+		// Should not happen, if it does, likely a bug.
+		logger.Printf("Unable to get remote IP(?!?!): %v", err)
```

Comment:
Should we exit here?

Reply:
Probably, I'll panic.

Reply:
... as it turns out, this breaks the test suite because none of the faked HTTP requests from tests set RemoteIP.